packages/headless/src/features/commerce/recommendation/recommendation-state.ts

```diff
@@ -0,0 +1,22 @@
+import {CommerceAPIErrorStatusResponse} from '../../../api/commerce/commerce-api-error-response';
```

Comment:
Ultimately we will want to support 0-* concurrent recommendation interfaces in the Commerce engine state so that multiple carousels can be displayed on the same page without having to initialize multiple engines (I think we want to avoid that at all cost).

This implies that the recommendation state should probably be defined as a set (e.g., `Record<string, RecommendationSlice>`). The record key could potentially be the recommendation's slot ID (assuming that it is intended as a unique identifier).

When calling the Recommendation controller builder, the recommendation interface would be "registered" in the state, and its state selector would point to `engine.state.recommendation[slotId]` (this is similar to what we do for facets, for instance).

The trickier part is pagination. Right now, the pagination state is "global" in the Commerce engine, which hasn't been an issue so far because we only ever display one product listing or search interface at a time. In order to support multiple concurrent paginated interfaces in the same engine state, however, the Recommendation, ProductListing, and Search controllers should each have their own Pagination controller / state.

The approach proposed in https://coveord.atlassian.net/wiki/spaces/COM/pages/3907256341/Commerce+Headless+Routing is to have those "parent" controllers return builders for the sub-controller that they are likely to require.

Should we do this right now, or in an ulterior phase?

Reply:
I would propose we do it now, but in a second PR. Doing this now would not be a wasted effort since we will need to write that logic to support different state parts for a controller anyway.

Reply:
Aye! 👍